src/proto/sfs/mod.rs

```diff
@@ -0,0 +1,182 @@
+use {Status, Result, ucs2};
+use core::mem;
+
+pub const FILE_MODE_READ    : u64 = 0x0000000000000001;
+pub const FILE_MODE_WRITE   : u64 = 0x0000000000000002;
+pub const FILE_MODE_CREATE  : u64 = 0x8000000000000000;
+
+pub const FILE_READ_ONLY    : u64 = 0x0000000000000001;
+pub const FILE_HIDDEN       : u64 = 0x0000000000000002;
+pub const FILE_SYSTEM       : u64 = 0x0000000000000004;
+pub const FILE_RESERVED     : u64 = 0x0000000000000008;
+pub const FILE_DIRECTORY    : u64 = 0x0000000000000010;
+pub const FILE_ARCHIVE      : u64 = 0x0000000000000020;
+pub const FILE_VALID_ATTR   : u64 = 0x0000000000000037;
+
+#[repr(C)]
+pub struct File {
+    revision: u64,
+    open: extern "C" fn(this: &mut File, new_handle: &mut usize, filename: *const u16, open_mode: u64, attributes: u64) -> Status,
+    close: extern "C" fn(this: &mut File) -> Status,
+    delete: extern "C" fn(this: &mut File) -> Status,
+    read: extern "C" fn(this: &mut File, buffer_size: &mut usize, buffer: *mut u8) -> Status,
+    write: extern "C" fn(this: &mut File, buffer_size: &mut usize, buffer: *const u8) -> Status,
+    get_position: extern "C" fn(this: &mut File, position: &mut u64) -> Status,
+    set_position: extern "C" fn(this: &mut File, position: u64) -> Status,
+    get_info: usize,
+    set_info: usize,
+    flush: extern "C" fn(this: &mut File) -> Status,
+}
+
+#[repr(C)]
+pub struct SimpleFileSystem {
+    revision: u64,
+    open_volume: extern "C" fn(this: &mut SimpleFileSystem, root: &mut usize) -> Status, 
+}
+
+impl File {
+    /// Try to open a file relative to this file/directory.
+    ///
+    /// # Arguments
+    /// * `filename`    Path of file to open, relative to this File
+    /// * `open_mode`   The mode to open the file with. Valid
+    ///     combinations are FILE_MODE_READ, FILE_MODE_READ | FILE_MODE_WRITE and
+    ///     FILE_MODE_READ | FILE_MODE_WRITE | FILE_MODE_CREATE
+    /// * `attributes`  Only valid when FILE_MODE_CREATE is used as a mode
+    /// 
+    /// # Errors
+    /// * `uefi::Status::NotFound`          Could not find file
+    /// * `uefi::Status::NoMedia`           The device has no media
+    /// * `uefi::Status::MediaChanged`      The device has a different medium in it
+    /// * `uefi::Status::DeviceError`       The device reported an error
+    /// * `uefi::Status::VolumeCorrupted`   The filesystem structures are corrupted
+    /// * `uefi::Status::WriteProtected`    Write/Create attempted on readonly file
+    /// * `uefi::Status::AccessDenied`      The service denied access to the file
+    /// * `uefi::Status::OutOfResources`    Not enough resources to open file
+    /// * `uefi::Status::VolumeFull`        The volume is full
+    pub fn open(&mut self, filename: &str, open_mode: u64, attributes: u64) -> Result<&mut File> {
+        const BUF_SIZE : usize = 128;
```

Comment:
Max file name size on FAT32 is 255 UCS-2 characters, we should support at least that many bytes. Should also document that this is the limit, and mention that the `InvalidParameter` error can also be returned.